src/main/java/uk/nhs/hee/trainee/details/api/DspCredentialResource.java

```diff
@@ -0,0 +1,267 @@
+      }
+      idTokenHint = jwtService.generatePlacementToken(placement.get());
+      scope = "issue.TestCredential"; //TODO set up issue.Placement
+
+    } else if (credentialType.equalsIgnoreCase("programmemembership")) {
+      Optional<ProgrammeMembership> programmeMembership
+          = programmeMembershipService.getProgrammeMembershipForTrainee(traineeTisId, tisId);
+      if (programmeMembership.isEmpty()) {
+        log.warn("Unable to find programme membership with ID {} for trainee with ID {}",
+            tisId, traineeTisId);
+        return ResponseEntity.unprocessableEntity().build();
+      }
+      idTokenHint = jwtService.generateProgrammeMembershipToken(programmeMembership.get());
+      scope = "issue.TestCredential"; //TODO set up issue.ProgrammeMembership
+    }
+
+    String nonce = UUID.randomUUID().toString();
+    MultiValueMap<String, String> bodyPair = new LinkedMultiValueMap<>();
+    bodyPair.add("client_id", this.clientId);
+    bodyPair.add("client_secret", this.clientSecret);
+    bodyPair.add("redirect_uri", this.redirectUri);
+    bodyPair.add("scope", scope);
+    bodyPair.add("id_token_hint", idTokenHint);
+    bodyPair.add("nonce", nonce);
+    bodyPair.add("state", STATE);
+
+    HttpHeaders headers = new HttpHeaders();
+    headers.setAccept(Collections.singletonList(MediaType.APPLICATION_JSON));
+    headers.setContentType(MediaType.APPLICATION_FORM_URLENCODED);
+    HttpEntity<MultiValueMap<String, String>> parRequest = new HttpEntity<>(bodyPair, headers);
+    RestTemplate restTemplate = this.restTemplateBuilder
+        .setConnectTimeout(Duration.ofMillis(CONNECT_TIMEOUT))
+        .setReadTimeout(Duration.ofMillis(READ_TIMEOUT))
+        .build();
+    URI parUri = URI.create(this.parEndpoint);
+
+    ResponseEntity<ParResponse> parResponseEntity = restTemplate.postForEntity(parUri, parRequest,
+        ParResponse.class);
+    if (parResponseEntity.getStatusCode() == HttpStatus.CREATED) {
+      ParResponse parResponse = parResponseEntity.getBody();
+      if (parResponse != null) {
+        String location = String.format("%s?client_id=%s&request_uri=%s", issueEndpoint, clientId,
+            parResponse.getRequestUri());
+        return ResponseEntity.created(URI.create(location)).build();
+      }
+    }
+    return ResponseEntity.internalServerError().build();
+  }
+
+  /**
+   * Get the credential content from a (recently) issued credential.
+   *
+   * @param code  The auth code
+   * @param state The state used in the original PAR call
+   * @return The credential contents JSON or Bad Request if the code or state are invalid / expired,
+   *         or InternalServerError if the gateway response is invalid.
+   */
+  @GetMapping("/payload")
+  public ResponseEntity<String> getCredentialPayload(@RequestParam String code,
+      @RequestParam String state) {
+    //I wonder if we need @RequestHeader(HttpHeaders.AUTHORIZATION) String token just to be safer?
```

Comment:
We only need the token if we have to retrieve data from it, we can always assume that the token is verified by API Gateway prior to the endpoint being hit.